p2p: add `DifferenceFormatter` fuzz target and invariant check

frankomosh opened the pull request:
Adds a fuzz test for the [`DifferenceFormatter`](https://github.com/bitcoin/bitcoin/blob/e3f416dbf7633b2fb19c933e5508bd231cc7e9cf/src/blockencodings.h#L22-L42) (used in [`BlockTransactionsRequest`](https://github.com/bitcoin/bitcoin/blob/master/src/blockencodings.h#L44-L54), [BIP 152](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki)). The DifferenceFormatter class implements differential encoding for compact block transactions (BIP 152). This PR ensures that its strictly-monotonic property is maintained. It complements the tests in [`blocktransactionsrequest_deserialize`](https://github.com/bitcoin/bitcoin/blob/9703b7e6d563ea58f83a6eca819562365404f4ab/src/test/fuzz/deserialize.cpp#L314). 

Additionally, there's an added invariant check after GETBLOCKTXN deserialization in `net_processing.cpp`.

maflcko reviewed (commented):
(left my previous comment as in-line review comments)

frankomosh commented:
> Thanks, but the value of purely LLM generated "vibe" pull requests is limited, because:
> 
> * The "author" does not understand the changes and can not reply to code review feedback.
> * There are more than 300 pull requests (most written by real humans) waiting for review.
> * Incentive to review purely LLM generated changes is limited, because the feedback likely does not help to train a real human planning to contribute in the future, but is likely only passed on to an LLM api.
> 
> Also, specifically here:
> 
> * The added P2P assert is not fuzz-only related (it will be run in production as well)
> * `src/test/fuzz/CMakeLists.txt` isn't sorted
> * `src/test/fuzz/difference_formatter.cpp` is missing the copyright header (I guess the llm was trained on removing copyright headers)
> * The fuzz target `FUZZ_TARGET(difference_formatter)` doesn't add any new code coverage (this point was also already explained to you in your previous attempt at adding a fuzz target)
> * For a trivial serialization roundrip, you can just use the existing `./src/test/fuzz/deserialize.cpp` framework.
> * The CI (lint) doesn't pass.
> 
> So I'll close this for now. My recommendation for the future would be to create your first pull request fully by yourself. This will hopefully ensure that you understand the changes yourself and can reply to feedback.
> 
> ( Generally, I think it is fine to use LLMs, if the author would have written the same code almost exactly the same way without the LLM, and also fully understands it, and also can claim copyright on it. But this is probably a meta discussion. )

I will work to resolve the issues you raised as soon as I can please

frankomosh commented on src/net_processing.cpp:
```diff
@@ -4135,6 +4135,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // DifferenceFormatter should guarantee this property during deserialization
+        for (size_t i = 1; i < req.indexes.size(); ++i) {
+            assert(req.indexes[i] > req.indexes[i-1]);
+        }
```
Ok. I'll change this

brunoerg replied:
Stiil needs to update the PR title.

frankomosh replied:
@brunoerg The PR title was already updated yesterday, to "fuzz: add target for DifferenceFormatter and p2p invariant check" (changed from the previous version that mentioned net_processing.cpp). Please let me know if you’d still like to see any wording tweaks?

frankomosh commented:
dropped the serialization roundrip as @maflcko  suggests, they can be done elsewhere (./src/test/fuzz/deserialize.cpp).

frankomosh committed:
p2p: add assertion for BlockTransactionsRequest indexes

Adds Assume() check in net_processing after deserialization and validate DifferenceFormatter Class invariant.

frankomosh commented on src/test/fuzz/difference_formatter.cpp:
```diff
@@ -0,0 +1,57 @@
+#include <vector>
+
+namespace {
+struct TestDifferenceFormatterType {
```
I think this is more cleaner and simpler to follow. thanks

frankomosh committed:
fuzz: add a target for DifferenceFormatter Class

Add fuzz test to verify that arbitrary input successfully deserialized
by DifferenceFormatter will maintain the sorted-without-duplicates invariant.

frankomosh commented:
Updated `fuzz/difference_formatter.cpp` to directly use `BlockTransactionsRequest`

dergoegge commented on src/net_processing.cpp:
```diff
@@ -4135,6 +4135,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        // Verify differential encoding invariant: indexes must be strictly increasing
+        // DifferenceFormatter should guarantee this property during deserialization
+        for (size_t i = 1; i < req.indexes.size(); ++i) {
+            assert(req.indexes[i] > req.indexes[i-1]);
```
```suggestion
            Assume(req.indexes[i] > req.indexes[i-1]);
```

Just so the assertion isn't active in release builds

frankomosh replied:
yes. thanks

dergoegge commented on src/test/fuzz/difference_formatter.cpp:
```diff
@@ -0,0 +1,39 @@
+
+namespace {
+void VerifyDifferenceFormatterInvariants(const std::vector<uint16_t>& indexes) {
+    // Core invariant: strictly monotonic increasing (no duplicates allowed)
+    for (size_t i = 1; i < indexes.size(); ++i) {
+        assert(indexes[i] > indexes[i-1]); // Must be STRICTLY greater
+    }
```
The two comments are saying the same thing, I'd suggest to just keep one

dergoegge commented on src/test/fuzz/difference_formatter.cpp:
```diff
@@ -0,0 +1,39 @@
+        assert(test_container.blockhash == block_hash);
+
+        // If deserialization succeeded, verify the DifferenceFormatter invariants
+        VerifyDifferenceFormatterInvariants(test_container.indexes);
```
There's only one invariant we're checking, perhaps we don't need the function?

frankomosh replied:
Had only kept it for readability, but yes we can definately do without it

dergoegge reviewed (approved):
Code review ACK 65a10fc3c52ea09a4794345bcf607dff908c783a

achow101 commented on src/net_processing.cpp:
```diff
@@ -4135,6 +4135,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         vRecv >> req;
+        // Verify differential encoding invariant: indexes must be strictly increasing
+        // DifferenceFormatter should guarantee this property during deserialization
+        for (size_t i = 1; i < req.indexes.size(); ++i) {
+            Assume(req.indexes[i] > req.indexes[i-1]);
+        }
```
In 65a10fc3c52ea09a4794345bcf607dff908c783a "p2p: add assertion for BlockTransactionsRequest indexes"

Can checking this be done in deserialization directly, rather than iterating all of the indexes again? Since this is part of compact block relay, we want to avoid doing as much unnecessary work as possible.

frankomosh replied:
Thanks for the review @achow101 

I agree it could be done directly in deserialization. The reason it’s in net_processing is that it acts as a boundary for untrusted peer input, so having the invariant explicit here serves as a defense-in-depth check (similar to other message level sanity checks here).
The extra loop is only executed after we already know the message is GETBLOCKTXN and after successful deserialization, so it’s a one-time O(n) pass per compact block request. Given the bounded size of these requests, I think the cost is negligible compared to the overall block relay path?

dergoegge replied:
Since this is using Assume, the loop shouldn't exist in the release binaries (i.e the compiler optimises it out), so the overhead doesn't exist.
